### ISSUE TITLE
Bug when using the plugin

### ISSUE THREAD

**Issue by pwnsdx** [waysact/webpack-subresource-integrity#15]
Hello,

Thanks you for this plugin. I'm actually getting a very weird bug while using your plugin.

Here is the error:

Error parsing 'integrity' attribute ('*-*-*-CHUNK-SRI-HASH-1-*-*-*'). The specified hash algorithm must be one of 'sha256', 'sha384', or 'sha512'.
requireEnsure @ main.3bf4ea4….js:8036
(anonymous) @ category.3bf4ea4….js:44
__webpack_require__ @ main.3bf4ea4….js:7395
webpackJsonpCallback @ main.3bf4ea4….js:22
(anonymous) @ category.3bf4ea4….js:1

Error parsing 'integrity' attribute ('*-*-*-CHUNK-SRI-HASH-16-*-*-*'). The specified hash algorithm must be one of 'sha256', 'sha384', or 'sha512'.

[the number change often, sometime it's like this for all chunks]

**Also, I noticed that sometime (when running webpack), the stuff you put in the JS is not even present (but the integrity tags etc on js/css files are correctly there in the HTML)**

The causing code:

All requires seems to be concerned but here is an example of a require I do
```
require(['./includes/js/MenusSlider'], (MenusSlider) => {
     return (new MenusSlider(defines.MenusSlider));
});
```

and the SRI plugin part in Webpack:
https://gist.github.com/pwnsdx/b73aa8baf8d11c4be3570f08cd84b1d3

Which is very big probably because of that bug...

I hope we can find a way to fix this issue and make this plugin better. If you need any informations feel free to ask me.

S.

**Comment by pwnsdx** [comment]
Also, I was wondering, is your plugin works well with plugins like [promise-loader](https://github.com/gaearon/promise-loader)?

Regards,
S

**Comment by jscheid** [comment]
Hi, thanks for your bug report. Could you create a test project that demonstrates the issue?

**Comment by jscheid** [comment]
And yes, it should work fine with promise-loader. We're using this plugin and promise-loader in the same project and haven't had any issues.

**Comment by jscheid** [comment]
If you don't find it easy to reproduce the issue in a project that you can share, perhaps it would be enough if you could send over your webpack config and your `package.json`.

**Comment by pwnsdx** [comment]
Hello,

I'm going to try to reproduce the issue with a fresh webpack config. If I'm unable I'll send you my webpack config and package.json. Do you have a PGP public key?

Regards,

**Comment by Lokaltog** [comment]
I'm receiving this error as well. It neven happens on static builds, but happens when I use webpack-dev-server. Steps to reproduce:

1. Start webpack-dev-server
2. Edit and save a file (triggering a rebuild)
3. Refresh browser

Console error:

```manifest.bundle.js?v=fe3007:742
Error parsing 'integrity' attribute ('*-*-*-CHUNK-SRI-HASH-4-*-*-*'). The specified hash algorithm must be one of 'sha256', 'sha384', or 'sha512'.```

Versions:

webpack-dev-server 2.1.0-beta.11
webpack 2.1.0-beta.27
webpack-subresource-integrity 0.6.0

**Comment by Lokaltog** [comment]
I followed the stack trace and the error appears to be triggered if you use module splitting in webpack 2, combined with webpack-dev-server.

This works: `import foo from 'bar';`

This does not: `System.import('foo').then(...)`

**Comment by jscheid** [comment]
@Lokaltog interesting, that would make sense because of incremental rebuilds. It would also explain why I've never encountered the issue (we're using SRI only in production, not in development mode. And we're not using hot reloading in production either.)

@pwnsdx are you also using webpack-dev-server when you see the issue?

I will have to think a bit more about this, but my initial reaction is that we might not be able to support compatibility with webpack-dev-server. @Lokaltog out of curiosity, and assuming that you are using webpack-dev-server during development only, is there a reason you need SRI in development, or is it just that you don't use separate webpack configs for development and production?

At any rate, I'm treating this as a bug because at the very least the UX needs to be improved, perhaps with a better error message.

**Comment by pwnsdx** [comment]
Hi there,

Yes. I was using webpack-dev-server while having this bug. But I'm on Webpack 1, not 2. Going to try a production build to see if everything is okay, if yes then it might be the root cause.

Regards,
S.

**Comment by Lokaltog** [comment]
@jscheid I ended up disabling it for development. The only reason it was enabled in my dev setup to begin with was that I put it in the array of shared plugins instead of the production-specific plugins.

I can't imagine anyone needing this in development, so if this error is limited to webpack-dev-server and difficult to fix, I think a note about how to avoid this issue in the docs would be sufficient.

**Comment by pwnsdx** [comment]
Yeah, disabled it too, I was using it un dev mode because I wasn't aware that it would be problematic this way.

By the way, here is what I get when using "webpack" in production mode in the HTML files:

```
Html Webpack Plugin:
<pre>
  TypeError: Cannot read property 'integrity' of undefined
  
  - index.js:231 assetIntegrity
    [x-HTML]/[webpack-subresource-integrity]/index.js:231:43
  
  - Array.map
  
  - index.js:227 addIntegrity
    [x-HTML]/[webpack-subresource-integrity]/index.js:227:39
  
  - Array.forEach
  
  - index.js:225 Compilation.beforeHtmlGeneration
    [x-HTML]/[webpack-subresource-integrity]/index.js:225:21
  
  - Tapable.js:86 Compilation.applyPluginsAsyncWaterfall
    [x-HTML]/[tapable]/lib/Tapable.js:86:13
  
  - util.js:16 Compilation.tryCatcher
    [x-HTML]/[html-webpack-plugin]/[bluebird]/js/release/util.js:16:    23
  
  
  - index.js:627 
    [x-HTML]/[html-webpack-plugin]/index.js:627:12
  
  - index.js:120 
    [x-HTML]/[html-webpack-plugin]/index.js:120:16
  
  - util.js:16 tryCatcher
    [x-HTML]/[html-webpack-plugin]/[bluebird]/js/release/util.js:16:    23
  
  - promise.js:510 Promise._settlePromiseFromHandler
    [x-HTML]/[html-webpack-plugin]/[bluebird]/js/release/promise.js:    510:31
  
  - promise.js:567 Promise._settlePromise
    [x-HTML]/[html-webpack-plugin]/[bluebird]/js/release/promise.js:    567:18
  
  - promise.js:612 Promise._settlePromise0
    [x-HTML]/[html-webpack-plugin]/[bluebird]/js/release/promise.js:    612:10
  
  - promise.js:691 Promise._settlePromises
    [x-HTML]/[html-webpack-plugin]/[bluebird]/js/release/promise.js:    691:18
  
  - async.js:138 Async._drainQueue
    [x-HTML]/[html-webpack-plugin]/[bluebird]/js/release/async.js:13    8:16
  
  - async.js:148 Async._drainQueues
    [x-HTML]/[html-webpack-plugin]/[bluebird]/js/release/async.js:14    8:10
  
  - async.js:17 Immediate.Async.drainQueues
    [x-HTML]/[html-webpack-plugin]/[bluebird]/js/release/async.js:17    :14
  
</pre>
```

For the JS files, I get no traces of any hash, "integrity" or "sriHashes" in the code. Moving `new SriPlugin` before and after `new HtmlWebpackPlugin` changes nothing.

**Comment by pwnsdx** [comment]
Okay, I think I identified where the bug come from, it occur when using [extract-text-webpack-plugin](https://github.com/webpack/extract-text-webpack-plugin) (`extractCSS = new ExtractTextPlugin('[name].[hash].css')`)

**Comment by jscheid** [comment]
@pwnsdx can you see if it works better when using #16 ?

    npm install waysact/webpack-subresource-integrity#better-plugin-registration

It would also be interesting to know if ☝️ this works better together with webpack-dev-server, in which case we wouldn't have to add the warning to the README. Can you guys give that a try perhaps?

**Comment by Lokaltog** [comment]
I've tested #16 and it appears to build fine with WDS with no errors. But I'm receiving a couple of new warnings in the console:

```2 client?cd17:90 webpack-subresource-integrity: chunks loaded by HMR are unprotected.```

**Comment by Lokaltog** [comment]
I see now that these warnings are actually emitted by the fix in #16 so that makes sense. But it makes it annoying to use in development since they're emitted in the console, so I think I'd still add a note in the README clarifying the plugin behavior when used with WDS.

**Comment by jscheid** [comment]
Do you think it would work if we limited it to a single warning per session?
-- 
Sent from a mobile device

**Comment by pwnsdx** [comment]
Hello,

Still the same issue :/

After investigating: it looks like beforeHtmlGeneration -> hwpSrcRelativeOutputPath is trying to get "integrity" key for "../main.8908839fb89551fac84b.js" but the real file is "main.8908839fb89551fac84b.js"

The only part in my code where I have ../ is this part:

```
...
        new HtmlWebpackPlugin({
            filename: '../index.html',
            template: 'src/index.pug',
            chunks: ['main', 'index']
        }),
        new HtmlWebpackPlugin({
            filename: '../product.html',
            template: 'src/product.pug',
            chunks: ['main', 'product']
        }),
        new HtmlWebpackPlugin({
            filename: '../category.html',
            template: 'src/category.pug',
            chunks: ['main', 'category']
        }),
...
```

and this:

```
...
    'css-loader?' + JSON.stringify({
        safe: true,
        calc: false,
        minimize: IS_PRODUCTION,
        discardComments: {
            removeAll: IS_PRODUCTION
        },
        publicPath: '../'
    }),
...
```

So I added in `beforeHtmlGeneration` function, this:

`src = path.parse(src).base;`

So no more errors but nothing in the JS/html anymore.

**Comment by jscheid** [comment]
@pwnsdx I don't think it's the same issue, this looks like a different issue from what you originally reported. Can you send the `output` section of your webpack config please? Or even better, your whole webpack config.

**Comment by jscheid** [comment]
@pwnsdx Never mind, I can reproduce like this. I'll take a closer look later today.

**Comment by jscheid** [comment]
@pwnsdx can you try with https://github.com/waysact/webpack-subresource-integrity/pull/17 ?

    npm install waysact/webpack-subresource-integrity#fix-hwp-asset-src

**Comment by pwnsdx** [comment]
@jschied Glad that you were able to reproduce and sorry for the response time.

So I tried #17 and I can confirm it now works great. Thanks you for resolving this issue.

**Comment by jscheid** [comment]
That's great news @pwnsdx! I'll do a few more tweaks and wait for some more feedback from other people, planning to roll a new release in the coming week.

**Comment by pwnsdx** [comment]
Awesome! Can't wait to use the new release :)

Thanks again for this plugin,

Cheers!

**Comment by jscheid** [comment]
@Lokaltog I'd like to keep the warning, to make sure nobody uses it with HMR in production without knowing about the limitations. (If HMR in production sounds crazy, the [docs actually suggest it as a use case](https://github.com/webpack/docs/wiki/hot-module-replacement-with-webpack#what-can-i-do-with-it).)

However I've made sure it's only ever printed once, and also I am going to [make it easy to disable the plugin in development mode](https://github.com/waysact/webpack-subresource-integrity/tree/new-config-syntax#webpack-configuration-example) in which case of course no warning will be printed.

I'll also [add a few words about it to the README](https://github.com/waysact/webpack-subresource-integrity/tree/new-config-syntax#hot-module-reloading).

What do you think?

**Comment by jscheid** [comment]
The fix for this was released in v0.7.0.

**Comment by pwnsdx** [comment]
Awesome! Thanks you! :+1:
